docs/user-defined-type-functions.md

```diff
@@ -0,0 +1,230 @@
+
+<details><summary>Expand for full type API reference.</summary>
+
+#### `types`
+
+| Instance Attributes | Type | Description |
+| ------------- | ------------- | ------------- |
+| `unknown` | `type` | an immutable instance of the built-in type `unknown` |
+| `never` | `type` | an immutable instance of the built-in type `never` |
+| `any` | `type` | an immutable instance of the built-in type `any` |
+| `boolean` | `type` | an immutable instance of the built-in type `boolean` |
+| `number` | `type` | an immutable instance of the built-in type `number` |
+| `string` | `type` | an immutable instance of the built-in type `string` |
+
+| Instance Methods | Return Type | Description |
+| ------------- | ------------- | ------------- |
+| `__eq(arg: type)` | `boolean` | overrides the == operator to return true if self is syntactically equal to arg |
+| `is(arg: string)` | `boolean` | returns true if self has the same tag as the argument. List of available tags: "nil", "unknown", "never", "any", "boolean", "number", "string", "singleton", "negation", "union", "intersection", "table", "function", "class"  |
+
+| Static Methods | Return Type | Description |
+| ------------- | ------------- | ------------- |
+| `singleton(arg: string \| boolean \| nil)` | `type` | returns an immutable instance of the argument as a singleton type |
+| `negationof(arg: type)` | `type` | returns an immutable instance of the parameter as a negated type; the parameter cannot be a table or function type |
+| `unionof(...: type)` | `type` | returns an immutable instance of an union type of the arguments; requires at least 2 parameters to be provided |
+| `intersectionof(...: type)` | `type` | returns an immutable instance of an intersection type of the arguments; requires at least 2 parameters to be provided |
+| `newtable(props: {[type]: type}?, indexer: {key: type, value: type}?, metatable: type?)` | `type` | returns a mutable instance of a table type; the keys of the table's property must be a singleton type; if provided the metatable parameter, this table becomes a metatable |
+| `newfunction(parameters: {pack: {type}?, variadic: type?}, returns: {pack: {type}?, variadic: type?})` | `type` | returns a mutable instance of a `function` type |
+| `copy(arg: type)` | `type` | returns a deep copy of the given `type` |
+
+#### Negation
+
+| Instance Methods | Type | Description |
+| ------------- | ------------- | ------------- |
+| `inner()` | `type` | returns the `type` being negated |
+
+#### Singleton
+
+| Instance Methods | Return Type | Description |
+| ------------- | ------------- | ------------- |
+| `value()` | `string \| boolean` | returns the string value of the  singleton |
```

Comment:
```suggestion
| `value()` | `string \| boolean` | returns the string value of the singleton |
```

Reply:
There is a bit of imbalance between `singleton(nil)` used to create a `nil` type and `nil` type not actually being a singleton.

Reply:
I think we can reasonably make `value()` work for `nil` too.